trx_file_memmap/cpp/src/trx.cpp

```diff
@@ -0,0 +1,361 @@
+#include "trx.h"
+#include <fstream>
+#include <typeinfo>
+#include <errno.h>
+#include <algorithm>
+#define SYSERROR() errno
+
+//#define ZIP_DD_SIG 0x08074b50
+//#define ZIP_CD_SIG 0x06054b50
+using namespace Eigen;
+using namespace std;
+
+std::string get_base(const std::string &delimeter, const std::string &str)
+{
+	std::string token;
+
+	if (str.rfind(delimeter) + 1 < str.length())
+	{
+		token = str.substr(str.rfind(delimeter) + 1);
+	}
+	else
+	{
+		token = str;
+	}
+	return token;
+}
+
+std::string get_ext(const std::string &str)
+{
+	std::string ext = "";
+	std::string delimeter = ".";
+
+	if (str.rfind(delimeter) + 1 < str.length())
+	{
+		ext = str.substr(str.rfind(delimeter));
+	}
+	return ext;
+}
+
+namespace trxmmap
+{
+	// TODO: check if there's a better way
+	int _sizeof_dtype(std::string dtype)
+	{
+		// TODO: make dtypes enum??
+		auto it = std::find(dtypes.begin(), dtypes.end(), dtype);
+		int index = 0;
+		if (it != dtypes.end())
+		{
+			index = std::distance(dtypes.begin(), it);
+		}
+
+		switch (index)
+		{
+		case 1:
+			return 1;
+		case 2:
+			return sizeof(uint8_t);
+		case 3:
+			return sizeof(uint16_t);
+		case 4:
+			return sizeof(uint32_t);
+		case 5:
+			return sizeof(uint64_t);
+		case 6:
+			return sizeof(int8_t);
+		case 7:
+			return sizeof(int16_t);
+		case 8:
+			return sizeof(int32_t);
+		case 9:
+			return sizeof(int64_t);
+		case 10:
+			return sizeof(float);
+		case 11:
+			return sizeof(double);
+		default:
+			return sizeof(half); // setting this as default for now but a better solution is needed
+		}
+	}
+
+	std::string _get_dtype(std::string dtype)
+	{
+		char dt = dtype.back();
+		switch (dt)
+		{
+		case 'b':
+			return "bit";
+		case 'h':
+			return "uint8";
+		case 't':
+			return "uint16";
+		case 'j':
+			return "uint32";
+		case 'm':
+			return "uint64";
+		case 'a':
+			return "int8";
+		case 's':
+			return "int16";
+		case 'i':
+			return "int32";
+		case 'l':
+			return "int64";
+		case 'f':
+			return "float32";
+		case 'd':
+			return "float64";
+		default:
+			return "float16"; // setting this as default for now but a better solution is needed
+		}
+	}
+	std::tuple<std::string, int, std::string> _split_ext_with_dimensionality(const std::string filename)
+	{
+
+		// TODO: won't work on windows and not validating OS type
+		std::string base = get_base("/", filename);
```

Comment:
Since we're using C++11, I wonder if setting a constant so we could use it in the project would be better, not requiring parameterizing the delimiter.

e.g.

```
#if defined(WIN32) || defined(_WIN32) 
#define PATH_SEPARATOR "\\" 
#else 
#define PATH_SEPARATOR "/" 
#endif 
```

Reply:
That's a good idea! I was trying to defer to built-in functions for handling path difference, but use of compiler directives would be useful either way and this is a good way to get things working on Windows for now.